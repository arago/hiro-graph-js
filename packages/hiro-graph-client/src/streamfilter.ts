export class StreamFilter {
  private value = '';

  constructor(value?: string) {
    if (value) {
      this.value = value;
    }
  }

  private reduce(text: (string | StreamFilter)[]) {
    return text.reduce((acc, t) => acc + `(${t.toString()})`, '');
  }

  or(...text: (string | StreamFilter)[]) {
    const v = this.reduce(text);

    this.value += `|${v}`;

    return this;
  }

  and(...text: (string | StreamFilter)[]) {
    const v = this.reduce(text);

    this.value += `&${v}`;

    return this;
  }

  toString() {
    return `${this.value}`;
  }
}
