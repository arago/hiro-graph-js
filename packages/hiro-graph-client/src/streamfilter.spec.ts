import { StreamFilter } from './streamfilter';

test('StreamFilter snapshots', () => {
  const tests = [
    new StreamFilter().and('a = b', 'd = e'),
    new StreamFilter().or('a = b', 'd = e'),
    new StreamFilter().and('a = b', new StreamFilter('d = e')),
    new StreamFilter().and(
      'a = b',
      new StreamFilter().or('e = f').and('d = c', 'g = f'),
    ),
  ];

  tests.map((t) => expect(t.toString()).toMatchSnapshot());
});

test('StreamFilter functions', () => {
  expect(new StreamFilter().and('a', 'b').toString()).toBe('&(a)(b)');
  expect(new StreamFilter().or('a', 'b').toString()).toBe('|(a)(b)');
});

test('StreamFilter nested', () => {
  expect(
    new StreamFilter().and('a', new StreamFilter().and('b', 'c')).toString(),
  ).toBe('&(a)(&(b)(c))');
  expect(
    new StreamFilter().or('a', new StreamFilter().or('b', 'c')).toString(),
  ).toBe('|(a)(|(b)(c))');
});
