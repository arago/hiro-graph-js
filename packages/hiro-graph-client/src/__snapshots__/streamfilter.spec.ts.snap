// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamFilter snapshots 1`] = `"&(a = b)(d = e)"`;

exports[`StreamFilter snapshots 2`] = `"|(a = b)(d = e)"`;

exports[`StreamFilter snapshots 3`] = `"&(a = b)(d = e)"`;

exports[`StreamFilter snapshots 4`] = `"&(a = b)(|(e = f)&(d = c)(g = f))"`;
